File: java/test/Fib.java
// The system under test: a tiny Fibonacci library.
//
// Every omni port ships this same library, with the same behaviour and the
// same error messages, and tests it with the same spec file
// (spec/fib.json).

import com.voxgig.omni.Util;
import java.util.LinkedHashMap;
import java.util.Map;

public final class Fib {

  private Fib() {}

  /** Validate a Fibonacci index. The spec matches on these messages. */
  public static long fibindex(Object val) {
    if (!Util.isnum(val)) {
      throw new IllegalArgumentException("fib: not a number: " + Util.stringify(val));
    }

    double num = ((Number) val).doubleValue();

    if (num != Math.rint(num)) {
      throw new IllegalArgumentException("fib: non-integer index: " + Util.stringify(val));
    }

    if (num < 0) {
      throw new IllegalArgumentException("fib: negative index: " + Util.stringify(val));
    }

    return (long) num;
  }

  /** The nth Fibonacci number: fib(0)=0, fib(1)=1. */
  public static long fib(Object n) {
    return fibnum(fibindex(n));
  }

  public static long fibnum(long index) {
    if (0 == index) {
      return 0;
    }

    long prev = 0;
    long cur = 1;

    for (long step = 1; step < index; step++) {
      long next = prev + cur;
      prev = cur;
      cur = next;
    }

    return cur;
  }

  /** The first n Fibonacci numbers. */
  public static Object fibseq(Object n) {
    long count = fibindex(n);
    java.util.List<Object> out = new java.util.ArrayList<>();
    for (long index = 0; index < count; index++) {
      out.add(fibnum(index));
    }
    return out;
  }

  /** The Fibonacci numbers from index start to index end, inclusive. */
  public static Object fibrange(Object start, Object end) {
    long from = fibindex(start);
    long to = fibindex(end);
    java.util.List<Object> out = new java.util.ArrayList<>();
    for (long index = from; index <= to; index++) {
      out.add(fibnum(index));
    }
    return out;
  }

  /**
   * A description of the nth Fibonacci number. `prev` is null at index 0,
   * which exercises the runner's null handling.
   */
  public static Object fibinfo(Object n) {
    long index = fibindex(n);
    long val = fibnum(index);

    Map<String, Object> out = new LinkedHashMap<>();
    out.put("n", index);
    out.put("val", val);
    out.put("even", 0 == val % 2);
    out.put("prev", 0 == index ? null : fibnum(index - 1));
    out.put("label", "fib(" + Util.numstr(index) + ")=" + Util.numstr(val));

    return out;
  }
}

File: lua/test/fib.lua
-- The system under test: a tiny Fibonacci library.
--
-- Every omni port ships this same library, with the same behaviour and the
-- same error messages, and tests it with the same spec file
-- (spec/fib.json).

local u = require('util')

local M = {}

--- Validate a Fibonacci index. The spec matches on these messages.
function M.fibindex(val)
  if not u.isnum(val) then
    error('fib: not a number: ' .. u.stringify(val), 0)
  end

  if val ~= math.floor(val) then
    error('fib: non-integer index: ' .. u.stringify(val), 0)
  end

  if 0 > val then
    error('fib: negative index: ' .. u.stringify(val), 0)
  end

  return math.floor(val)
end

--- The nth Fibonacci number: fib(0)=0, fib(1)=1.
function M.fibnum(index)
  if 0 == index then
    return 0
  end

  local prev = 0
  local cur = 1

  for _ = 1, index - 1 do
    prev, cur = cur, prev + cur
  end

  return cur
end

function M.fib(val)
  return M.fibnum(M.fibindex(val))
end

--- The first n Fibonacci numbers.
function M.fibseq(val)
  local count = M.fibindex(val)
  local out = u.list({})
  for index = 0, count - 1 do
    out[#out + 1] = M.fibnum(index)
  end
  return out
end

--- The Fibonacci numbers from index start to index end, inclusive.
function M.fibrange(start, stop)
  local from = M.fibindex(start)
  local to = M.fibindex(stop)
  local out = u.list({})
  for index = from, to do
    out[#out + 1] = M.fibnum(index)
  end
  return out
end

--- A description of the nth Fibonacci number. `prev` is null at index 0,
--- which exercises the runner's null handling.
function M.fibinfo(val)
  local index = M.fibindex(val)
  local num = M.fibnum(index)

  return u.map({
    n = index,
    val = num,
    even = 0 == num % 2,
    prev = 0 == index and u.NULL or M.fibnum(index - 1),
    label = 'fib(' .. u.numstr(index) .. ')=' .. u.numstr(num),
  })
end

return M
